### PR TITLE
add docstrings to adjoint solver

### PR DESCRIPTION
**Pull request by mochen4**
Add docstrings to ``objective.py`` and ``optimization_problem.py``

### CODE REVIEW

File: python/adjoint/objective.py [hunk 1]
```diff
@@ -284,11 +293,30 @@ def __call__(self):
 
 
 class FourierFields(ObjectiveQuantity):
+    """A differentiable frequency-dependent Fourier Fields (dft_fields)
+    Attributes:
```
Comment (oskooi) [inline python/adjoint/objective.py:297]: Replace "Attributes:" with "Args:" and separate it with the one-line summary above it with a line break following the style guide: https://google.github.io/styleguide/pyguide.html#s3.8.3-functions-and-methods.

File: python/adjoint/objective.py [hunk 2]
```diff
@@ -284,11 +293,30 @@ def __call__(self):
...
+        volume: the volume over which the Fourier Fields are calculated. Due to an unsolved bug,
```
Comment (oskooi) [inline python/adjoint/objective.py:298]: "Fields" should be lowercase. Change "unsolved" to "unresolved."

File: python/adjoint/objective.py [hunk 3]
```diff
@@ -284,11 +293,30 @@ def __call__(self):
...
+          the size must not be zero in at least one direction.
+        component: field component (e.g. mp.Ex, mp.Hz, etc.)
+        yee_grid: if True, the Fourier fields evaluated at corresponding Yee grid point
```
Comment (oskooi) [inline python/adjoint/objective.py:301]: ...at *the* corresponding Yee grid point...

File: python/adjoint/objective.py [hunk 4]
```diff
@@ -284,11 +293,30 @@ def __call__(self):
...
+          are returned; otherwise, interpolated fields at the center of each voxel
+          are returned
+        decimation_factor: an integer so that the DFT fields are updated at every
```
Comment (oskooi) [inline python/adjoint/objective.py:304]: An integer used to specify the number of timesteps between updates of the DFT fields.

File: python/adjoint/objective.py [hunk 5]
```diff
@@ -385,6 +420,24 @@ def __call__(self):
 
 
 class Near2FarFields(ObjectiveQuantity):
+    """A differentiable near2far field transformation
+    Attributes:
```
Comment (oskooi) [inline python/adjoint/objective.py:424]: Args:

File: python/adjoint/optimization_problem.py [hunk 6]
```diff
@@ -21,6 +21,27 @@ class OptimizationProblem:
     calculation) and optionally its gradient (adjoint calculation).
     This is done by the __call__ method.
 
+    Attributes:
```
Comment (oskooi) [inline python/adjoint/optimization_problem.py:24]: Args:

**Review by oskooi — CHANGES_REQUESTED** [review]

File: python/adjoint/objective.py (hunk 1 above)
Comment (oskooi) [reply python/adjoint/objective.py:297]: `Attributes` is used for classes and `Args` for functions.

For the `Attributes` section of the docstrings for the `EigenmodeCoefficient` class, we would need to list its *public* members, almost anything with a `self.` prefix defined in the constructor excluding `self._` which are generally used for private members.

We should also add a separate `Args` section to the constructor. The argument descriptions in the constructor docstrings should hopefully be different than the attributes description in the class docstring. The attributes description should be a general summary whereas the argument description is less so.

Comment (mochen4) [reply python/adjoint/objective.py:312]: I added separate ``Args`` sections to the constructors. I tried to make the attribute description more concise and general, and argument description more detailed and elaborated. But they still seem a little redundant to me.

File: python/adjoint/optimization_problem.py [hunk 7]
```diff
@@ -123,7 +166,29 @@ def __call__(
...
+              and gradients will then be evaluated at this configuration of design variables.
+            need_value: whether forward evaluations are needed. Default is True.
+            need_gradient: whether adjoint and gradients evaluatiosn are needed. Default is True.
+            beta: the strength of projection of rho_vector. Default to None.
+
+        Returns:
+            A tuple (f0, gradient) where:
+            f0 is the list of objective functions values when design variables
+              are set to rho_vector
+            gradient is the lists of gradients of objective functions with respect to
```
Comment (stevengj) [inline python/adjoint/optimization_problem.py:183]: Clarify that `gradient` is a list (over objective functions) of lists (over design regions) of 2d arrays (design vars x frequencies) of derivatives.   If there is only a single objective function, the outer 1-element list is replaced by just that element, and similarly if there is only one design region, while if there is only 1 frequency then the innermost array is squeezed to a 1d array.  (So, for example, if you have only a single objective function, a single design region, and a single frequency, then `gradient` is simply a 1d array of the derivatives.)

File: python/adjoint/optimization_problem.py [hunk 8]
```diff
@@ -21,6 +21,26 @@ class OptimizationProblem:
     calculation) and optionally its gradient (adjoint calculation).
     This is done by the __call__ method.
 
+    Attributes:
+        sim: the corresponding Meep `Simulation` object of the problem
+        objective_functions: list of objective functions on objective_arguments
```
Comment (stevengj) [inline python/adjoint/optimization_problem.py:26]:
```suggestion
        objective_functions: list of differentiable functions (callable objects) whose arguments are given by objective_arguments
```

**Pull request merged**